kext/FileNVRAM/FileNVRAM.cpp

```diff
@@ -115,49 +115,115 @@ bool FileNVRAM::start(IOService *provider)
     if(!dict) return false;
     setPropertyTable(dict);
     
-    
     if(bootnvram)
     {
         copyEntryProperties(NULL, bootnvram);
         bootnvram->detachFromParent(root, gIODTPlane);
+        // We restored the nvram from the bootloader. register the service
+        finalizeInit();
+        // waiting for IOBSD to determine if is safe to sync..
     }
     
     if(mReadOnly)
     {
         // we assume that the bootloader has done its job
         // i.e. populate /chosen/nvram
-        mInitComplete = true;
         mSafeToSync = true;
-        registerNVRAM();
+        if(!mInitComplete)
+        {
+            finalizeInit();
+        }
         return true;
     }
     else
     {
-        IOTimerEventSource* mTimer = IOTimerEventSource::timerEventSource(this, timeoutOccurred);
-        
-        if(mTimer)
+        // waiting for a client to appear before watching for IOBSD
+        if(VERSION_MAJOR >10)
         {
-            getWorkLoop()->addEventSource( mTimer);
-            mTimer->setTimeoutMS(50); // callback isn't being setup right, causes a panic
-            mSafeToSync = false;
+            // added in 10.6
+            disk_notifier_start_ = addMatchingNotification(gIOMatchedNotification,
+                                                           serviceMatching("IOMediaBSDClient"),
+                                                           FileNVRAM::DiskFound_callback,
+                                                           this, nullptr, 0);
         }
         else
         {
-            earlyInit = true;
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+            // deprecated in 10.6
+            disk_notifier_start_ = addNotification(gIOMatchedNotification,
+                                                   serviceMatching("IOMediaBSDClient"),
+                                                   ((IOServiceNotificationHandler)&(FileNVRAM::DiskFound_callback)),
+                                                   this, nullptr, 0);
+#pragma clang diagnostic pop
+        }
+        
+        
+        if(disk_notifier_start_ == nullptr)
+        {
+            IOLog("FileNVRAM: disk_notifier_start_ == nullptr\n");
+            return false;
         }
     }
     
-    // We don't have initial nvram data from the bootloader, or we couldn't schedule a
-    // timer to read in the nvram file, so start up immediately.
-    if(earlyInit == true)
-    {
-        mSafeToSync = true;
-        registerNVRAM();
-    }
+    return true;
+}
+
+bool FileNVRAM::DiskFound_callback(void *target, void *ref, IOService *service, IONotifier *notifier)
+{
+    FileNVRAM *self = (FileNVRAM *)target;
     
-    mInitComplete = true;
+    notifier->remove();
+    notifier = NULL;
     
-    return true;
+    if(OSDictionary *matching = IOService::resourceMatching( "IOBSD" ))
```

Comment:
A question (hope to not bore you :-) ).
I'm testing, and I'm going to reintroduce the old code w/o making additional commits for now, but I wondering why to not set the FileNVRAM personality this way:

`<key>IOResourceMatch</key>
			<string>IOBSD</string>`

instead of efi-runtime. Any downside doing that?
Thanks.